src/plugins/vis_type_table/public/utils/use/use_ui_state.ts

```diff
@@ -0,0 +1,139 @@
+
+import { ColumnWidthData, TableVisUiState, TableVisUseUiStateProps } from '../../types';
+
+const defaultSort = {
+  columnIndex: null,
+  direction: null,
+};
+
+export const useUiState = (
+  uiState: IInterpreterRenderHandlers['uiState']
+): TableVisUseUiStateProps => {
+  const [sort, setSortState] = useState<TableVisUiState['sort']>(
+    uiState?.get('vis.params.sort') || defaultSort
+  );
+
+  const [columnsWidth, setColumnsWidthState] = useState<TableVisUiState['colWidth']>(
+    uiState?.get('vis.params.colWidth') || []
+  );
+
+  const uiStateValues = useRef<{
+    columnsWidth: ColumnWidthData[];
+    sort: TableVisUiState['sort'];
+    /**
+     * Property to filter out chnages, which were done internally via local state.
+     */
+    pendingUpdate: boolean;
+  }>({
+    columnsWidth: uiState?.get('vis.params.colWidth'),
+    sort: uiState?.get('vis.params.sort'),
+    pendingUpdate: false,
+  });
+
+  const setSort = useCallback(
+    (s: TableVisUiState['sort'] = defaultSort) => {
+      setSortState(s || defaultSort);
+
+      uiStateValues.current.sort = s;
+      uiStateValues.current.pendingUpdate = true;
+
+      /**
+       * Since the visualize app state is listening for uiState changes,
+       * it synchronously re-renders an editor frame.
+       * Setting new uiState values in the new event loop task,
+       * helps to udpate visualization frame firstly and not to block the rendering flow
+       */
+      setTimeout(() => {
+        uiState?.set('vis.params.sort', s);
+        uiStateValues.current.pendingUpdate = false;
+      });
+    },
+    [uiState]
+  );
+
+  const setColumnsWidth = useCallback(
+    (col: ColumnWidthData) => {
+      setColumnsWidthState((prevState) => {
+        const updated = [...prevState];
+        const idx = prevState.findIndex((c) => c.colIndex === col.colIndex);
+
+        if (idx >= 0) {
+          updated[idx] = col;
+        } else {
+          updated.push(col);
+        }
+
+        uiStateValues.current.columnsWidth = updated;
+        uiStateValues.current.pendingUpdate = true;
+
+        /**
+         * Since the visualize app state is listening for uiState changes,
+         * it synchronously re-renders an editor frame.
+         * Setting new uiState values in the new event loop task,
+         * helps to udpate visualization frame firstly and not to block the rendering flow
+         */
+        setTimeout(() => {
+          uiState?.set('vis.params.colWidth', updated);
+          uiStateValues.current.pendingUpdate = false;
+        });
+        return updated;
+      });
+    },
+    [uiState]
+  );
+
+  useEffect(() => {
+    /**
+     * Debounce is in place since there are couple of synchronous updates of the uiState,
+     * which are also handled synchronously.
+     */
+    const updateOnChange = debounce(() => {
+      const { vis } = uiState?.getChanges();
+
+      // skip uiState updates if there are pending internal state updates
```

Comment:
nit: you can move the return at the beginning of the function and then get the vis